posix/fdtable.c

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/bits/safecrt.h>
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/unistd.h>
+#include <openenclave/enclave.h>
+#include <openenclave/internal/posix/device.h>
+#include <openenclave/internal/posix/fdtable.h>
+#include <openenclave/internal/posix/raise.h>
+#include <openenclave/internal/print.h>
+#include <openenclave/internal/thread.h>
+#include <openenclave/internal/trace.h>
+#include <openenclave/internal/utils.h>
+
+/*
+**==============================================================================
+**
+** Define the table of file-descriptor entries:
+**
+**==============================================================================
+*/
+
+/* The table allocation grows in multiples of the chunk size. */
+#define TABLE_CHUNK_SIZE 1024
+
+typedef oe_device_t* entry_t;
+
+static entry_t* _table;
+static size_t _table_size;
+static oe_spinlock_t _lock = OE_SPINLOCK_INITIALIZER;
+static bool _installed_atexit_handler;
+
+static void _atexit_handler(void)
+{
+    oe_free(_table);
+}
+
+static int _resize_table(size_t new_size)
+{
+    int ret = -1;
+
+    /* Install the atexit handler on the first call. */
+    if (!_installed_atexit_handler)
+    {
+        oe_atexit(_atexit_handler);
+        _installed_atexit_handler = true;
+    }
+
+    /* The fdtable cannot be bigger than the maximum int file descriptor. */
+    if (new_size > OE_INT_MAX)
+        goto done;
+
+    /* Round the new capacity up to the next multiple of the chunk size. */
+    new_size = oe_round_up_to_multiple(new_size, TABLE_CHUNK_SIZE);
+
+    if (new_size > OE_INT_MAX)
+        new_size = OE_INT_MAX;
```

Comment:
Unnecessary given the prior check on line 54.

Reply:
It is necessary, because of the rounding performed on the previous line could increase **new_size** beyond **OE_INT_MAX**. Note that someone could change the value of TABLE_CHUNK_SIZE in the future so that **OE_INT_MAX** is no longer a multiple of **TABLE_CHUNK_SIZE**.